agent/consul/gateway_locator.go

```diff
@@ -0,0 +1,284 @@
+	primaryGateways []string // WAN addrs
+	localGateways   []string // LAN addrs
+
+	// primaryMeshGatewayDiscoveredAddresses is the current fallback addresses
+	// for the mesh gateways in the primary datacenter.
+	primaryMeshGatewayDiscoveredAddresses     []string
+	primaryMeshGatewayDiscoveredAddressesLock sync.Mutex
+
+	// This will be closed the FIRST time we get some gateways populated
+	primaryGatewaysReadyCh chan struct{}
+}
+
+// PrimaryMeshGatewayAddressesReadyCh returns a channel that will be closed
+// when federation state replication ships back at least one primary mesh
+// gateway (not via fallback config).
+func (g *GatewayLocator) PrimaryMeshGatewayAddressesReadyCh() <-chan struct{} {
+	return g.primaryGatewaysReadyCh
+}
+
+// PickGateway returns the address for a gateway suitable for reaching the
+// provided datacenter.
+func (g *GatewayLocator) PickGateway(dc string) string {
+	item := g.pickGateway(dc == g.primaryDatacenter)
+	g.logger.Trace("picking gateway for transit", "gateway", item, "source_datacenter", g.datacenter, "dest_datacenter", dc)
+	return item
+}
+
+func (g *GatewayLocator) pickGateway(primary bool) string {
+	addrs := g.listGateways(primary)
+	return getRandomItem(addrs)
+}
+
+func (g *GatewayLocator) listGateways(primary bool) []string {
+	g.gatewaysLock.Lock()
+	defer g.gatewaysLock.Unlock()
+
+	var addrs []string
+	if primary {
+		addrs = g.primaryGateways
+	} else {
+		addrs = g.localGateways
+	}
+
+	if primary && len(addrs) == 0 {
+		addrs = g.PrimaryGatewayFallbackAddresses()
+	}
+
+	return addrs
+}
+
+// RefreshPrimaryGatewayFallbackAddresses is used to update the list of current
+// fallback addresses for locating mesh gateways in the primary datacenter.
+func (g *GatewayLocator) RefreshPrimaryGatewayFallbackAddresses(addrs []string) (int, error) {
+	s := g // TODO
+	sort.Strings(addrs)
+
+	s.primaryMeshGatewayDiscoveredAddressesLock.Lock()
+	defer s.primaryMeshGatewayDiscoveredAddressesLock.Unlock()
+
+	if lib.StringSliceEqual(addrs, s.primaryMeshGatewayDiscoveredAddresses) {
+		return 0, nil
+	}
+
+	s.primaryMeshGatewayDiscoveredAddresses = addrs
+
+	s.logger.Info("updated fallback list of primary mesh gateways", "mesh_gateways", addrs)
+
+	return len(addrs), nil
+}
+
+// PrimaryGatewayFallbackAddresses returns the current set of discovered
+// fallback addresses for the mesh gateways in the primary datacenter.
+func (g *GatewayLocator) PrimaryGatewayFallbackAddresses() []string {
+	s := g // TODO
+	s.primaryMeshGatewayDiscoveredAddressesLock.Lock()
+	defer s.primaryMeshGatewayDiscoveredAddressesLock.Unlock()
+
+	out := make([]string, len(s.primaryMeshGatewayDiscoveredAddresses))
+	copy(out, s.primaryMeshGatewayDiscoveredAddresses)
+	return out
+}
+
+func getRandomItem(items []string) string {
+	switch len(items) {
+	case 0:
+		return ""
+	case 1:
+		return items[0]
+	default:
+		idx := int(rand.Int31n(int32(len(items))))
+		return items[idx]
+	}
+}
+
+type serverDelegate interface {
```

Comment:
🤔 why decouple this component from the server code but then keep it in the same package? I don't mind particularly but we often use this pattern to lift a component into a different package without circular dependencies/simpler testing.

I'm guessing by the `s := g // TODO` aliases that this was once built into the server directly but now isn't for a good reason though - perhaps that's worth documenting here?

Reply:
It was for testing, to avoid having to spin up an entire server just to test the light logic in here.